### PR TITLE
Add DSpace ingest

### PR DESCRIPTION
**Pull request by hakbailey**
#### What does this PR do?

Adds the ability to ingest from a DSpace source. Uses pre-approved
mapping from METS XML to TIMDEX record metadata. Changes include:

- Add a dspace record Struct to map XML unmarshaller
- Add a dspace generator to parse each record
- Add tests for each metadata field, with a sample export file from
  DSpace@MIT
- Add a dspace opton to ingester configuration

Not that there are still a few metadata fields to be mapped, for which I'm waiting on sample records. I'll create a new ticket for those so they don't get forgotten.

#### How can a reviewer manually see the effects of these changes?

`mario ingest -c json -t dspace fixtures/dspace_samples.xml` to see the JSON output or `mario ingest -t dspace fixtures/dspace_samples.xml` to ingest into a local elasticsearch instance on the default port.

#### What are the relevant tickets?

- https://mitlibraries.atlassian.net/browse/DISCO-158

#### Requires Full Reindexing of all Sources?
NO (will require indexing of new source only)

#### Includes new or updated dependencies?
NO

#### Todo:
- [X] Tests
- [ ] Documentation
- [ ] Stakeholder approval

### CODE REVIEW

File: pkg/generator/dspace.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	// Language field
+	var languages []string
+	for _, l := range mods.Languages {
+		languages = append(languages, l.LanguageTerm)
+	}
+	r.Language = skipEmpty(languages)
+
+	// Links field
+	for _, i := range mods.Identifiers {
+		if i.Type == "uri" {
```
Comment (JPrevost) [inline pkg/generator/dspace.go:107]: Are there link types that are not `url`? If so are they potentially useful even if they aren't the digital object?

File: pkg/generator/dspace.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+				Kind: "Digital object URI",
+				URL:  i.Value,
+				Text: "Digital object URI",
+			}
+			r.Links = append(r.Links, uri)
+		}
+	}
+
+	// PublicationDate field
+	r.PublicationDate = mods.OriginInfo.DateIssued
```
Comment (JPrevost) [inline pkg/generator/dspace.go:118]: No change requested, but one day we'll need to come up with a solid plan to normalize dates across systems to allow for useful faceting by date. We have a ticket for this so no action is required now but wanted to put it on the radar for future work.

**Review by JPrevost — COMMENTED** [review]
I'd like confirmation on the `uri` question and potentially a function name changed to reflect it ended up not being DC and confirmation if we can skip the `rulesfile` if we aren't using it. Otherwise this looks and works great. 👍🏻

File: pkg/generator/dspace.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+			if se.Name.Local == "record" {
+				processDCRecord(se, decoder, out)
+			}
+		}
+	}
+	close(out)
+}
+
+// processDCRecord handles the mapping from OAI_DC to Record.
+func processDCRecord(se xml.StartElement, decoder *xml.Decoder, out chan record.Record) {
```
[the remark this answers is not in the document]
Comment (hakbailey) [reply pkg/generator/dspace.go:53]: Definitely, will fix.

File: pkg/generator/dspace.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+	r := record.Record{}
+
+	// ContentType field
+	r.ContentType = mods.Genre
+
+	// Contributor field
+	if len(mods.Contributors) > 0 {
+		for _, c := range mods.Contributors {
+			contrib := new(record.Contributor)
+			contrib.Kind = c.Role.RoleTerm.RoleName
```
[the remark this answers is not in the document]
Comment (hakbailey) [reply pkg/generator/dspace.go:68]: Yeah, I'm not sure what all the options are coming from DSpace but we could probably get that list, remove/rename, and dedupe. I'll create a Jira ticket.

File: pkg/generator/dspace.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+)
+
+type dspaceparser struct {
+	file io.Reader
+}
+
+// DspaceGenerator parses dspace oai_dc xml data.
+type DspaceGenerator struct {
+	Dspacefile io.Reader
+	rulesfile  string
```
[the remark this answers is not in the document]
Comment (hakbailey) [reply pkg/generator/dspace.go:19]: Yep, will do.

File: pkg/generator/dspace.go (hunk 1 above)
Comment (hakbailey) [reply pkg/generator/dspace.go:107]: We don't currently have any links mapped from the dspace metadata other than the item handle. My understanding based on conversations with the metadata folks (and a review of the DSpace metadata fields) is that there are not often other links in the metadata.

Comment (JPrevost) [reply pkg/generator/dspace.go:106]: lol `there are not often other links in the metadata`. All good. I'm curious if we should map anything else that's in there anyway with a type of `unknown` or something. We could then analyze that and see if the stakeholders are remembering accurately what's in the data. I'm good with doing nothing here based on your info from stakeholders for sure, but I'm also slightly nervous there may be some other class of links they are forgetting about.

**Review by JPrevost — APPROVED** [review]
Thanks for the changes. I'm excited to see more data flowing into timdex :shipit:

**Pull request merged**